src/extension.ts

```diff
@@ -383,22 +367,90 @@ async function terraformCommand(command: string, languageServerExec = true): Pro
 }
 
 function enabled(): boolean {
-  return config('terraform').get('languageServer.external', false);
+  return config('terraform').get('languageServer.enable', false);
 }
 
-async function migrateLegacySettings() {
-  if (config('terraform').has('languageServer.enabled')) {
-    try {
-      await config('terraform').update(
-        'languageServer',
-        { enabled: undefined, external: true },
-        vscode.ConfigurationTarget.Global,
+async function migrateLegacySettings(ctx: vscode.ExtensionContext) {
+  // User has asked not to check if settings need to be migrated, so return
+  if (ctx.globalState.get('terraform.disableSettingsMigration', false)) {
+    return;
+  }
+
+  // TODO: Short circuit here so we don't check for every activation forever
+
+  // If any of the following list needs to be migrated, ask user if they want
+  // to migrate. This is a blunt force approach, but we don't intend to keep
+  // checking this forever
+  const warnMigration = warnIfMigrate([
+    { section: 'terraform', name: 'languageServer.external' },
+    { section: 'terraform', name: 'languageServer.pathToBinary' },
+    { section: 'terraform-ls', name: 'rootModules' },
+    { section: 'terraform-ls', name: 'excludeRootModules' },
+    { section: 'terraform-ls', name: 'ignoreDirectoryNames' },
+    { section: 'terraform-ls', name: 'terraformExecPath' },
+    { section: 'terraform-ls', name: 'terraformExecTimeout' },
+    { section: 'terraform-ls', name: 'terraformLogFilePath' },
+    { section: 'terraform-ls', name: 'experimentalFeatures' },
+  ]);
+  if (warnMigration === false) {
+    return;
+  }
+
+  // Prompt the user if they want to migrate. If the choose no, then return
+  // and they are left to migrate the settings themselves.
+  // If they choose yes, then automatically migrate the settings
+  // Lastly user can be directed to our README for more information about this
+  const choice = await vscode.window.showInformationMessage(
+    'Terraform configuration settings have changed in the latest update',
+    {
+      detail: 'You can automatically migrate or ask to not be reminded again and migrate the settings yourself',
+      modal: true,
+    },
+    { title: 'Migrate' },
+    { title: 'More Info' },
+    { title: "Don't remind me again" },
+  );
+  if (choice === undefined) {
+    return;
+  }
+
+  switch (choice.title) {
+    case "Don't remind me again":
+      ctx.globalState.update('terraform.disableSettingsMigration', true);
+      return;
+    case 'More Info':
+      await vscode.commands.executeCommand(
+        'vscode.open',
+        vscode.Uri.parse('https://www.github.com/hashicorp/vscode-terraform#settings-migration'),
```

Comment:
This is very useful! The only downside I noticed is that given that it's a button in multiple-choice dialog, it means that the dialog disappears, so if - after reading the details - user decides to do the migration, they have to reload the IDE to bring the dialog back.

Is there any way we could keep the dialog in place in this context? e.g. could the question in the dialog be markdown with that link? 🤔

Reply:
> This is very useful! The only downside I noticed is that given that it's a button in multiple-choice dialog, it means that the dialog disappears, so if - after reading the details - user decides to do the migration, they have to reload the IDE to bring the dialog back.

Unfortunately, yes, the modal window disappears when a button is clicked. However the window will come back every time the user starts VS Code until all settings we define in lines 392-402 are changed. We could implement a timer like we had with the LS updater to prompt the user, but that seems like it would upset the user more than help.

> Is there any way we could keep the dialog in place in this context? e.g. could the question in the dialog be markdown with that link? 🤔

With a dialog, not that I know of. These message windows have to be cross platform, so there are limited things you can do.